Реализует перемещение автомобилей

maver1cus opened the pull request:
- добавляет класс Universe
- разделяет config игры на отдельные сущности
- создает utils и const
- добавляет метод передвижение Car

maver1cus committed:
feat: добавляет движение автомобилей
- добавляет класс Universe
- разделяет config игры на отдельные сущности
- создает utils и const
- добавляет метод передвижение Car

kicumkicum commented on src/engine/universe.js:
```diff
@@ -0,0 +1,30 @@
+export default class Universe {
```
Я не очень большой знаток английского и возможно по этому это слово я перевожу как универсальный. Потом я понимаю, что что-то не то и лезу в переводчик. Я такой не один. Когда кто-то будет читать твой код, он не сможет залезть тебе в голову, но нужно, чтобы он понял. Тем более, этим кем-то можешь быть и ты через пол года. 

При выборе имени стоит руководствоваться правилом Проще - лучше. Брать то, что лежит на поверхности. Например, альтернатива ускоренной это слово Мир world. Или space(но это уже не очень, потому что или пространство, или космос). Или система координат

maver1cus replied:
done

kicumkicum commented on src/engine/universe.js:
```diff
@@ -0,0 +1,30 @@
+    return coordsY >= 0 && coordsY <= this._size.y;
+  }
+
+  validateCoords(coords) {
```
В методе происходит изменение внешних (входных данных). Это не очень хорошая практика, потому что нельзя менять чужое.

maver1cus replied:
done

kicumkicum commented on src/engine/universe.js:
```diff
@@ -0,0 +1,30 @@
+export default class Universe {
```
Сейчас класс содержит размерность карты и набор методов-валидаторов. Больше он ничего не делает. Возникает вопрос, нужен ли он на самом деле или можно обойтись парой утилитарных функций и константой с размерностью карты?

maver1cus replied:
done

kicumkicum commented on src/engine/universe.js:
```diff
@@ -0,0 +1,30 @@
+  }
+
+
+  isValidCoordX(coordsX) {
```
Если метод не используется снаружи, то его лучше сделать приватным

maver1cus replied:
done

kicumkicum commented on src/engine/game.js:
```diff
@@ -1,14 +1,24 @@
   gameStep() {
-    console.log(`game step:`, this._config);
+    const direction = getDirection();
+    this._cars.forEach(car => car.moving(direction, this._universe));
```
Сейчас расклад такой, что за движение отвечает три сущности `Game`, `Car`, `Universe`. Логика размазалась по нескольким местам и с ней будет сложно работать (в силу человеческой ограниченности).

Попробуй сделать предположение текстом, без кода, как можно упростить реализацию перемещения объектов в игре. Где какую логику разместить.

kicumkicum commented on src/engine/universe.js:
```diff
@@ -0,0 +1,30 @@
+
+  validateCoords(coords) {
+    if (!this.isValidCoordX(coords.x)) {
+      coords.x = (coords.x < 0)
```
Зарядка для ума - как можно написать это, используя `Math.min/Math.max`?

maver1cus replied:
done

kicumkicum reviewed (changes requested):
Код хорошо написан, но есть слабые стороны у основной задачи. Один коммент призывает к обсуждению этой реализации

maver1cus committed:
feat: изменяет работу движение автомобилей
    - переименовывает класс Universe в World
    - добавляет в World массив с элементами мира и их позицию
    - создает функцию по пересчету World

kicumkicum commented on src/engine/world.js:
```diff
@@ -0,0 +1,72 @@
+    }
+  }
+
+  render() {
```
Метод ничего не рендерит. Он просто обновляет положение предметов

maver1cus replied:
done

kicumkicum commented on src/engine/world.js:
```diff
@@ -0,0 +1,72 @@
+    this._size = config.worldSize;
+    this._elementsPositions = new Map();
+    this._elements = [];
+    this.i = 0;
```
Вроде как не используется

maver1cus replied:
done

kicumkicum commented on src/utils.js:
```diff
@@ -0,0 +1,15 @@
+  return array[randomIndex];
+}
+
+export const getDirection = () => getRandomItemFromArray(DIRECTIONS);
```
Я бы порекомендовал объявлять функции полным синтаксисом. Так их проще считывать, когда быстро читаешь код. Чем меньше вариативности в визуальных паттернах, тем проще мозгу работать. Но это не жесткое правило, просто мое наблюдение. Это относится к вкусовщине.

maver1cus replied:
done

kicumkicum commented on src/utils.js:
```diff
@@ -0,0 +1,15 @@
+  return array[randomIndex];
+}
+
+export const getDirection = () => getRandomItemFromArray(DIRECTIONS);
```
`getDirection` уже контекстная функция. Ее лучше объявить ближе к месту использования

maver1cus replied:
done

kicumkicum commented on src/engine/world.js:
```diff
@@ -0,0 +1,72 @@
+        const speed = element.speed;
+        const direction = element.directionRide;
+        let coords = this._elementsPositions.get(element);
+        switch (direction) {
```
Рекомендую отбивать пустой строкой объявление переменных и код

maver1cus replied:
done

kicumkicum commented on src/engine/world.js:
```diff
@@ -0,0 +1,72 @@
+
+      this._elements.push(newElement);
+      this._elementsPositions.set(newElement, element.coords);
+
```
Лишняя пустая строка и нет `;` ниже

maver1cus replied:
done

kicumkicum commented on src/engine/item.js:
```diff
@@ -1,3 +1,4 @@
 export default class Item {
-
+  constructor() {
+  }
```
Пустой конструктор можно не писать. И даже нужно не писать

maver1cus replied:
done

kicumkicum reviewed (changes requested):
Все круто! Начинают всплывать взрослые вещи из мира ООП. Я их описал ниже. Сейчас их фиксить не нужно - лучше сделать позже, потому что явно они фичу не блокируют. Заведи, пожалуйста, по такске на каждую. Про инверсию зависимостей можно почитать статейки, чтобы понять о чем вообще речь

Необходимый рефакторинг в будущем:
- [x] Константные строки поместить в енумы. Строки отстой сами по себе, хотя бы потому что сложно искать вхождения
- [x] Инстанцирование объектов убрать за пределы класса (инверсия зависимостей)

maver1cus committed:
feat: добавляет и настроаивает eslint

maver1cus committed:
refactor: рефакторинг кода
- исправляет отступы между участками кода
- переносит функцию получения направляения движения из utils в статический метод Car
- удаляет неиспользоваемые переменные
- удаляет пустой конструктор из Item

maver1cus committed:
refactor: перемименовывает функцию

kicumkicum commented on src/utils.js:
```diff
@@ -0,0 +1,12 @@
+export const getRandomInteger = function(a = 0, b = 1) {
```
Возможно я не корректно выразил свою мысль. В целом стрелочные функции лучше обычных function. Но однострочные стрелочные функции лучше использовать, когда ты их передаешь в качестве колбека. А когда ты создаешь функцию и сохраняешь в переменную, то лучше использовать полный ее синтаксис

```js
const foo = () => {
  return 42;
};
```
